File: peanoware/java/peanoware/Resources.java
/*
 * Operator.java
 *
 * Created on 7 ottobre 2004, 1.40
 */

package peanoware;

import java.awt.Color;
import java.awt.Font;

/**
 *
 * @author  monica
 */
public class Resources {
    public static final String AND = "And";
    public static final String OR = "Or";
    public static final String IMP = "Imp";
    public static final String NEG = "Neg";
    public static final String FORALL = "Forall";
    public static final String EXISTS = "Exist";
    public static final String APPLY = "Apply";
    public static final String QUANT_SEP = ".";
    public static final Color HYP = Color.BLUE;
    public static final Color FROM = Color.BLUE;
    public static final Color TO = Color.RED;
    public static final Color ARC = new Color(0, 163, 0);
    public static final String[] BUILD_OP = {AND, OR, IMP, NEG};
    public static final String[] BUILD_PROP_VAR = {"A", "B", "C", "D"};
    public static final String[] BUILD_PRED_SYMBOL = {"P", "Q", "R", "S"};
    public static final int[] BUILD_PRED_ARITY = {1, 1, 2, 2};
    public static final String[] BUILD_VAR = {"x", "y", "z", "t"};
    public static final Font FONT_FORMULA = new Font("Serif", Font.PLAIN, 35);
    public static final String CHALLENGE = "Challenge";
    public static final String NEXT = "Next";
    public static final String BRAVO = "Well done!!";
    public static final String PEANO_IMAGE = "peano2.png";
    public static final String PRAWITZ_IMAGE = "Dag.png";
    public static final String NULL_MESSAGE = " ";
    public static final String TYPES_ERROR = "Objects of wrong types";
    public static final String TYPE_ERROR = "Object of wrong type";
    public static final String ERROR_ARITY = "Only one item is allowed!";
    public static final String ERROR_ATTACH = "You can't perform this attachment";
    public static final String ERROR_SPLIT = "Only one item to dissolve!";
    public static final String ERROR_SPACE = "Can't make it. Resize the window please!";
    public static final String ERROR_COPY ="Only one item to copy!";
    public static final String ERROR_IMPOSSIBLE ="The formula requires this variable";
    public static final String ERROR_AND ="Only one active hypothesis is allowed";
    public static final String ERROR_TWO = "This operation requires two elements";
    public static String COPY = "Copy";
    public static String SPLIT = "Dissolve";
    public static String[] COMMAND = {COPY, SPLIT, AND, OR, IMP, NEG, FORALL, EXISTS, APPLY};
    public static String FALSE = "\u22A5";
    public static String BOX = "\u25A1";
    public static String[] INFO = 
        { null, null, " \u2227 ", " \u2228 ", " \u21d2 ", "\u00ac", "\u2200", "\u2203", ""};
    public static int[] ARITY = {1, 1, 2, 2, 2, 1, 2, 2, 2};
    public static int[] PREC = {1, 1, 30, 20, 10, 40, 7, 7, 60};
    public static int[] PREC_LEFT = {1, 1, 29, 19, 9, 41, 8, 8, 60};
    public static int[] PREC_RIGHT = {1, 1, 31, 21, 11, 40, 8, 8, 60};
    public static Color BACKGROUND = Color.BLACK;
    public static Color FOREGROUND = Color.WHITE;
    public static Color MESSAGE_COLOR = Color.RED;
    public static Color BORDER_COLOR = Color.LIGHT_GRAY;
    public static Color HIGHLIGHT_COLOR = new Color(0, 163, 0);
    public static Color BOX_COLOR = new Color(0, 163, 0);
    public static boolean validTopDown = false;
    public static boolean validBottomUp = true;
    
    public static String getString(String s) {
        for (int i = 0; i < COMMAND.length; i++) {
            if (s.equals(COMMAND[i])) {
                return INFO[i];
            }
        }
        return null;
    }
    
    public static String getCommand(String s) {
        for (int i = 0; i < COMMAND.length; i++) {
            if (s.equals(INFO[i])) {
                return COMMAND[i];
            }
        }
        return null;
    }
    
    public static int getArity(String s) {
        for (int i = 0; i < COMMAND.length; i++) {
            if (s.equals(COMMAND[i])) {
                return ARITY[i];
            }
        }
        return -1;
    }
    public static int getPrec(String s) {
        for (int i = 0; i < COMMAND.length; i++) {
            if (s.equals(COMMAND[i])) {
                return PREC[i];
            }
        }
        return 60;
    }
    public static int getLeftPrec(String s) {
        for (int i = 0; i < COMMAND.length; i++) {
            if (s.equals(COMMAND[i])) {
                return PREC_LEFT[i];
            }
        }
        return 60;
    }
    public static int getRightPrec(String s) {
        for (int i = 0; i < COMMAND.length; i++) {
            if (s.equals(COMMAND[i])) {
                return PREC_RIGHT[i];
            }
        }
        return 60;
    }
}

File: peanoware/javascript/Resources.js
/**
 * Resources.js
 *
 * Translated from Java (Created on 7 ottobre 2004, 1.40)
 * Original author: monica
 */

import Rectangle from './Rectangle.js'

const Resources = {
  // String constants
  AND: "And",
  OR: "Or",
  IMP: "Imp",
  NEG: "Neg",
  FORALL: "Forall",
  EXISTS: "Exist",
  APPLY: "Apply",
  QUANT_SEP: ".",
  
  // Color constants (using CSS color strings)
  HYP: "rgb(0, 0, 700)", // Blue
  FROM: "rgb(0, 0, 255)", // Blue
  TO: "rgb(255, 0, 0)", // Red
  ARC: "rgb(0, 163, 0)",
  
  // Array constants
  BUILD_OP: ["And", "Or", "Imp", "Neg"],
  BUILD_PROP_VAR: ["A", "B", "C", "D"],
  BUILD_PRED_SYMBOL: ["P", "Q", "R", "S"],
  BUILD_PRED_ARITY: [1, 1, 2, 2],
  BUILD_VAR: ["x", "y", "z", "t"],
  
  // Font constant (CSS font string)
  FONT_FORMULA: "35px Serif",
  
  // Message constants
  CHALLENGE: "Challenge",
  NEXT: "Next",
  BRAVO: "Well done!!",
  PEANO_IMAGE: "peano2.png",
  PRAWITZ_IMAGE: "Dag.png",
  NULL_MESSAGE: " ",
  TYPES_ERROR: "Objects of wrong types",
  TYPE_ERROR: "Object of wrong type",
  ERROR_ARITY: "Only one item is allowed!",
  ERROR_ATTACH: "You can't perform this attachment",
  ERROR_SPLIT: "Only one item to dissolve!",
  ERROR_SPACE: "Can't make it. Resize the window please!",
  ERROR_COPY: "Only one item to copy!",
  ERROR_IMPOSSIBLE: "The formula requires this variable",
  ERROR_AND: "Only one active hypothesis is allowed",
  ERROR_TWO: "This operation requires two elements",
  
  // Command constants
  COPY: "Copy",
  SPLIT: "Dissolve",
  
  // Computed array
  get COMMAND() {
    return [this.COPY, this.SPLIT, this.AND, this.OR, this.IMP, this.NEG, this.FORALL, this.EXISTS, this.APPLY];
  },
  
  // Special characters
  FALSE: "\u22A5",
  BOX: "\u25A1",
  
  // Info array
  INFO: [null, null, " \u2227 ", " \u2228 ", " \u21d2 ", "\u00ac", "\u2200", "\u2203", ""],
  
  // Numeric arrays
  ARITY: [1, 1, 2, 2, 2, 1, 2, 2, 2],
  PREC: [1, 1, 30, 20, 10, 40, 7, 7, 60],
  PREC_LEFT: [1, 1, 29, 19, 9, 41, 8, 8, 60],
  PREC_RIGHT: [1, 1, 31, 21, 11, 40, 8, 8, 60],
  
  // More color constants
  BACKGROUND: "rgb(0, 0, 0)", // Black
  FOREGROUND: "rgb(255, 255, 255)", // White
  MESSAGE_COLOR: "rgb(255, 0, 0)", // Red
  BORDER_COLOR: "rgb(211, 211, 211)", // Light Gray
  HIGHLIGHT_COLOR: "rgb(0, 163, 0)",
  BOX_COLOR: "rgb(0, 163, 0)",
  
  // Boolean flags
  validTopDown: false,
  validBottomUp: true,
  
  // Methods
  getString(s) {
    const commands = this.COMMAND;
    for (let i = 0; i < commands.length; i++) {
      if (s === commands[i]) {
        return this.INFO[i];
      }
    }
    return null;
  },
  
  getCommand(s) {
    const commands = this.COMMAND;
    for (let i = 0; i < commands.length; i++) {
      if (s === this.INFO[i]) {
        return commands[i];
      }
    }
    return null;
  },
  
  getArity(s) {
    const commands = this.COMMAND;
    for (let i = 0; i < commands.length; i++) {
      if (s === commands[i]) {
        return this.ARITY[i];
      }
    }
    return -1;
  },
  
  getPrec(s) {
    const commands = this.COMMAND;
    for (let i = 0; i < commands.length; i++) {
      if (s === commands[i]) {
        return this.PREC[i];
      }
    }
    return 60;
  },
  
  getLeftPrec(s) {
    const commands = this.COMMAND;
    for (let i = 0; i < commands.length; i++) {
      if (s === commands[i]) {
        return this.PREC_LEFT[i];
      }
    }
    return 60;
  },
  
  getRightPrec(s) {
    const commands = this.COMMAND;
    for (let i = 0; i < commands.length; i++) {
      if (s === commands[i]) {
        return this.PREC_RIGHT[i];
      }
    }
    return 60;
  },
  getRectRange(text, font, ctx, start, stop) {
    const oldfont = ctx.font;
    ctx.font = font;
    const textBefore = text.slice(0, start);
    const textUpToChar = text.slice(0, start + stop);
    const charMetrics = ctx.measureText(textUpToChar);
    const beforeMetrics = ctx.measureText(textBefore);
    const charX = beforeMetrics.width; // X position of the third character
    const charWidth = charMetrics.width - charX;
    const charHeight = charMetrics.actualBoundingBoxAscent + charMetrics.actualBoundingBoxDescent;
    ctx.font = oldfont; 
    return new Rectangle(
        charX - charMetrics.actualBoundingBoxLeft,
        - charMetrics.actualBoundingBoxAscent,
        charWidth,
        charHeight);
  },
  getRect(text, font, ctx) {
    const oldfont = ctx.font;
    ctx.font = font;
    const fullMetrics = ctx.measureText(text);
    const fullWidth = fullMetrics.width;
    const fullHeight = fullMetrics.actualBoundingBoxAscent + fullMetrics.actualBoundingBoxDescent;
    ctx.font = oldfont; 
    return new Rectangle(
                - fullMetrics.actualBoundingBoxLeft,
                - fullMetrics.actualBoundingBoxAscent,
                fullWidth,
                fullHeight
            );
  }
};

// For ES6 module export (if needed)
export default Resources;

// For CommonJS export (if needed)
// module.exports = Resources;